### PR TITLE
Make docker image with assertions

### PR DESCRIPTION
**Pull request by freenerd**
# Issue

We should build another class of docker images, that includes debug symbols and assertions. This will allow us to debug problems happening on large-scale datasets.

Sadly locally osrm doesn't build for me in assertion mode. Need to dig into that. Error message in details below.

## Tasklist
 - [x] Make sure docker images build locally in `release`, `debug` and `assertions` mode
 - [x] Add `latest-assertions` build rule to Docker Cloud
 - [ ] Merge

<details>

```
Scanning dependencies of target EXTRACTOR
[ 24%] Building CXX object CMakeFiles/EXTRACTOR.dir/src/extractor/compressed_edge_container.cpp.o
[ 25%] Building CXX object CMakeFiles/EXTRACTOR.dir/src/extractor/edge_based_graph_factory.cpp.o
[ 25%] Building CXX object CMakeFiles/EXTRACTOR.dir/src/extractor/extraction_containers.cpp.o
c++: internal compiler error: Killed (program cc1plus)
Please submit a full bug report,
with preprocessed source if appropriate.
See <http://gcc.gnu.org/bugs.html> for instructions.
make[2]: *** [CMakeFiles/SERVER.dir/build.make:159: CMakeFiles/SERVER.dir/src/server/api/parameters_parser.cpp.o] Error 4
make[1]: *** [CMakeFiles/Makefile2:176: CMakeFiles/SERVER.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[ 27%] Building CXX object CMakeFiles/EXTRACTOR.dir/src/extractor/extractor.cpp.o
[ 27%] Building CXX object CMakeFiles/EXTRACTOR.dir/src/extractor/extractor_callbacks.cpp.o
[ 29%] Building CXX object CMakeFiles/EXTRACTOR.dir/src/extractor/graph_compressor.cpp.o
[ 29%] Building CXX object CMakeFiles/EXTRACTOR.dir/src/extractor/raster_source.cpp.o
[ 30%] Building CXX object CMakeFiles/EXTRACTOR.dir/src/extractor/restriction_map.cpp.o
[ 30%] Building CXX object CMakeFiles/EXTRACTOR.dir/src/extractor/restriction_parser.cpp.o
[ 32%] Building CXX object CMakeFiles/EXTRACTOR.dir/src/extractor/scripting_environment_lua.cpp.o
[ 32%] Building CXX object CMakeFiles/EXTRACTOR.dir/src/extractor/suffix_table.cpp.o
[ 33%] Building CXX object CMakeFiles/EXTRACTOR.dir/src/extractor/guidance/coordinate_extractor.cpp.o
[ 33%] Building CXX object CMakeFiles/EXTRACTOR.dir/src/extractor/guidance/intersection.cpp.o
[ 35%] Building CXX object CMakeFiles/EXTRACTOR.dir/src/extractor/guidance/intersection_generator.cpp.o
[ 35%] Building CXX object CMakeFiles/EXTRACTOR.dir/src/extractor/guidance/intersection_handler.cpp.o
[ 37%] Building CXX object CMakeFiles/EXTRACTOR.dir/src/extractor/guidance/intersection_normalizer.cpp.o
[ 37%] Building CXX object CMakeFiles/EXTRACTOR.dir/src/extractor/guidance/mergable_road_detector.cpp.o
[ 38%] Building CXX object CMakeFiles/EXTRACTOR.dir/src/extractor/guidance/motorway_handler.cpp.o
[ 38%] Building CXX object CMakeFiles/EXTRACTOR.dir/src/extractor/guidance/node_based_graph_walker.cpp.o
[ 40%] Building CXX object CMakeFiles/EXTRACTOR.dir/src/extractor/guidance/roundabout_handler.cpp.o
[ 40%] Building CXX object CMakeFiles/EXTRACTOR.dir/src/extractor/guidance/sliproad_handler.cpp.o
[ 41%] Building CXX object CMakeFiles/EXTRACTOR.dir/src/extractor/guidance/suppress_mode_handler.cpp.o
[ 41%] Building CXX object CMakeFiles/EXTRACTOR.dir/src/extractor/guidance/turn_analysis.cpp.o
[ 43%] Building CXX object CMakeFiles/EXTRACTOR.dir/src/extractor/guidance/turn_classification.cpp.o
[ 43%] Building CXX object CMakeFiles/EXTRACTOR.dir/src/extractor/guidance/turn_discovery.cpp.o
[ 45%] Building CXX object CMakeFiles/EXTRACTOR.dir/src/extractor/guidance/turn_handler.cpp.o
[ 45%] Building CXX object CMakeFiles/EXTRACTOR.dir/src/extractor/guidance/turn_lane_augmentation.cpp.o
[ 46%] Building CXX object CMakeFiles/EXTRACTOR.dir/src/extractor/guidance/turn_lane_data.cpp.o
[ 46%] Building CXX object CMakeFiles/EXTRACTOR.dir/src/extractor/guidance/turn_lane_handler.cpp.o
[ 48%] Building CXX object CMakeFiles/EXTRACTOR.dir/src/extractor/guidance/turn_lane_matcher.cpp.o
[ 48%] Built target EXTRACTOR
make: *** [Makefile:128: all] Error 2
```

### CODE REVIEW

**Comment by freenerd** [comment]
This is now working for me locally. I also refactored the Dockerfile so that when building locally, only the changes to osrm-backend are run and the OS dependencies and stxxl stay cached.

I have added the build rule to docker cloud, so when we merge we should get assertion images.

@danpat @daniel-j-h @TheMarex can one of you have a look and merge?

File: README.md [hunk 1]
```diff
@@ -69,6 +69,15 @@ In case Docker complains about not being able to connect to the Docker daemon ma
...
 
+We support the following images on Docker Cloud:
+
+Name | Description
+-----|------
+`latest` | `master` compiled with release flag
+`latest-assertions` | `master` compiled with with release flag, assertions enabled and debug symbols
+`latest-debug` | `master` compiled with debug flag
+`<tag>` | specific tag compiled with release flag
+`<tag>-debug` | specific tag compiled with debug flag
```
Comment (daniel-j-h) [inline README.md]: should we have a tag release+assertions, then, too?

File: docker/Dockerfile [hunk 2]
```diff
@@ -23,15 +17,21 @@ RUN NPROC=$(grep -c ^processor /proc/cpuinfo 2>/dev/null || 1) && \
...
+RUN NPROC=$(grep -c ^processor /proc/cpuinfo 2>/dev/null || 1) && \
     echo "Building OSRM ${DOCKER_TAG}" && \
-    cd /src && \
     git show --format="%H" | head -n1 > /opt/OSRM_GITSHA && \
     echo "Building OSRM gitsha $(cat /opt/OSRM_GITSHA)" && \
     mkdir build && \
     cd build && \
-    cmake -DCMAKE_BUILD_TYPE=${BUILD_TYPE} -DENABLE_LTO=On .. && \
+    case ${DOCKER_TAG} in *"-debug"*) BUILD_TYPE="Debug";; *) BUILD_TYPE="Release";; esac && \
+    case ${DOCKER_TAG} in *"-assertions"*) echo "Building with assertions" && cmake .. -DCMAKE_BUILD_TYPE=RelWithDebInfo -DENABLE_ASSERTIONS=ON -DBUILD_TOOLS=1 -DENABLE_LTO=On ;; *) echo "Building for ${BUILD_TYPE}" && cmake .. -DCMAKE_BUILD_TYPE=${BUILD_TYPE} -DENABLE_LTO=On ;; esac && \
```
Comment (daniel-j-h) [inline docker/Dockerfile]: can we have a single `cmake -DOpt0=Val0 -DOpt1=Val1` here and the `case` branches just setting the appropriate values for all options?

File: README.md (hunk 1 above)
Comment (freenerd) [inline README.md]: What would be the use case of that? Who would actually use that? Given that every build on Docker Cloud takes 1 hour (and they are currently not parallelized, so each commit is 3h), I'm hesitant to add more.

**Pull request merged**